<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.433">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>Bringing genomics to the field - 2&nbsp; Genomic selection of sources</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<link href="./summary.html" rel="next">
<link href="./intro.html" rel="prev">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>


</head>

<body class="nav-sidebar floating">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
      <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./genomics_informed_provenancing.html"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Genomic selection of sources</span></a></li></ol></nav>
      <a class="flex-grow-1" role="button" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
      </a>
      <button type="button" class="btn quarto-search-button" aria-label="" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal sidebar-navigation floating overflow-auto">
    <div class="pt-lg-2 mt-2 text-left sidebar-header">
    <div class="sidebar-title mb-0 py-0">
      <a href="./">Bringing genomics to the field</a> 
    </div>
      </div>
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Preface</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./intro.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Introduction</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./genomics_informed_provenancing.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Genomic selection of sources</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./summary.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Summary</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./references.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">References</span></a>
  </div>
</li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#why-gd-and-gl" id="toc-why-gd-and-gl" class="nav-link active" data-scroll-target="#why-gd-and-gl"><span class="header-section-number">2.1</span> Why GD and GL?</a></li>
  <li><a href="#setting-up" id="toc-setting-up" class="nav-link" data-scroll-target="#setting-up"><span class="header-section-number">2.2</span> Setting up</a>
  <ul class="collapse">
  <li><a href="#import-annotations-from-snpeff" id="toc-import-annotations-from-snpeff" class="nav-link" data-scroll-target="#import-annotations-from-snpeff"><span class="header-section-number">2.2.1</span> Import annotations from SnpEff</a></li>
  <li><a href="#allele-probabilities-and-frequencies" id="toc-allele-probabilities-and-frequencies" class="nav-link" data-scroll-target="#allele-probabilities-and-frequencies"><span class="header-section-number">2.2.2</span> Allele probabilities and frequencies</a></li>
  </ul></li>
  <li><a href="#functions-to-optimize-selection" id="toc-functions-to-optimize-selection" class="nav-link" data-scroll-target="#functions-to-optimize-selection"><span class="header-section-number">2.3</span> Functions to optimize selection</a>
  <ul class="collapse">
  <li><a href="#select-regions-to-select-sources" id="toc-select-regions-to-select-sources" class="nav-link" data-scroll-target="#select-regions-to-select-sources"><span class="header-section-number">2.3.1</span> Select regions to select sources</a></li>
  <li><a href="#apply-the-function-to-get-optimal-source-combinations" id="toc-apply-the-function-to-get-optimal-source-combinations" class="nav-link" data-scroll-target="#apply-the-function-to-get-optimal-source-combinations"><span class="header-section-number">2.3.2</span> Apply the function to get optimal source combinations</a></li>
  </ul></li>
  <li><a href="#factors-affecting-source-selection" id="toc-factors-affecting-source-selection" class="nav-link" data-scroll-target="#factors-affecting-source-selection"><span class="header-section-number">2.4</span> Factors affecting source selection</a></li>
  <li><a href="#saving-data-for-downstream-analysis" id="toc-saving-data-for-downstream-analysis" class="nav-link" data-scroll-target="#saving-data-for-downstream-analysis"><span class="header-section-number">2.5</span> Saving data for downstream analysis</a></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Genomic selection of sources</span></h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  

</header>

<section id="why-gd-and-gl" class="level2" data-number="2.1">
<h2 data-number="2.1" class="anchored" data-anchor-id="why-gd-and-gl"><span class="header-section-number">2.1</span> Why GD and GL?</h2>
</section>
<section id="setting-up" class="level2" data-number="2.2">
<h2 data-number="2.2" class="anchored" data-anchor-id="setting-up"><span class="header-section-number">2.2</span> Setting up</h2>
<p>Read in the necessary data for the source optimization. SnpEff v5.1 (Cingolani et al., 2012) was used to annotate genetic variants to functional class based on Norway spruce genome annotation.</p>
<section id="import-annotations-from-snpeff" class="level3" data-number="2.2.1">
<h3 data-number="2.2.1" class="anchored" data-anchor-id="import-annotations-from-snpeff"><span class="header-section-number">2.2.1</span> Import annotations from SnpEff</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a>list <span class="ot">&lt;-</span> <span class="fu">list.files</span>(<span class="at">path =</span> <span class="st">"../../datashare/Spruce/exome_capture/WES_mapping/Annotations/ref_Pglauca/VCF_split_files"</span>, </span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a>                   <span class="at">pattern =</span> <span class="st">"Red_Spruce_intersect_poly_"</span>, </span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a>                   <span class="at">recursive=</span><span class="cn">TRUE</span>, <span class="at">full.names =</span> T)</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a><span class="co"># genes &lt;- lapply(list[1], function(x) read.table(x, nrow = 100000)) # originally run for testing</span></span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a>genes <span class="ot">&lt;-</span> <span class="fu">lapply</span>(list[<span class="dv">1</span>], <span class="cf">function</span>(x) <span class="fu">read.table</span>(x))</span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a>category <span class="ot">&lt;-</span> <span class="fu">lapply</span>(genes, <span class="cf">function</span>(x) <span class="fu">unlist</span>(<span class="fu">lapply</span>(<span class="fu">strsplit</span>(<span class="fu">as.character</span>(x[,<span class="dv">8</span>]), <span class="at">split =</span> <span class="st">"|"</span>, <span class="at">fixed =</span> T), <span class="cf">function</span>(y) y[<span class="dv">2</span>])))</span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true" tabindex="-1"></a>TAB <span class="ot">&lt;-</span> genes[<span class="dv">1</span><span class="sc">:</span><span class="dv">2</span>]</span>
<span id="cb1-10"><a href="#cb1-10" aria-hidden="true" tabindex="-1"></a>TAB <span class="ot">&lt;-</span> <span class="fu">do.call</span>(rbind, TAB)</span>
<span id="cb1-11"><a href="#cb1-11" aria-hidden="true" tabindex="-1"></a>category <span class="ot">&lt;-</span> <span class="fu">do.call</span>(c, category)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><em>Read in the meta data for the samples</em></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Info samples</span></span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a>names <span class="ot">&lt;-</span> <span class="fu">unlist</span>(<span class="fu">lapply</span>(<span class="fu">strsplit</span>(<span class="fu">unlist</span>(<span class="fu">strsplit</span>(<span class="fu">as.character</span>(<span class="fu">read.table</span>(<span class="st">"all_bam.list"</span>)[,<span class="dv">1</span>]), <span class="at">split =</span> <span class="st">"_rmdup.sorted.bam"</span>)), <span class="at">split =</span> <span class="st">"./"</span>), <span class="cf">function</span>(x) x[<span class="dv">2</span>]))</span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a>pops <span class="ot">&lt;-</span> <span class="fu">unlist</span>(<span class="fu">lapply</span>(<span class="fu">strsplit</span>(names, <span class="at">split=</span><span class="st">"_"</span>), <span class="cf">function</span>(x) x[<span class="dv">1</span>]))</span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a>info_inds <span class="ot">&lt;-</span> <span class="fu">read.table</span>(<span class="st">"./Info_samples_revised.txt"</span>, <span class="at">header=</span>T)</span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a>info_inds <span class="ot">&lt;-</span> info_inds[<span class="fu">match</span>(<span class="fu">as.character</span>(names), <span class="fu">as.character</span>(info_inds<span class="sc">$</span>Family)),]</span>
<span id="cb2-7"><a href="#cb2-7" aria-hidden="true" tabindex="-1"></a>info_pops <span class="ot">&lt;-</span> info_inds[<span class="sc">!</span><span class="fu">duplicated</span>(info_inds<span class="sc">$</span>Site),<span class="sc">-</span><span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">3</span>,<span class="dv">9</span>,<span class="dv">10</span>)]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="allele-probabilities-and-frequencies" class="level3" data-number="2.2.2">
<h3 data-number="2.2.2" class="anchored" data-anchor-id="allele-probabilities-and-frequencies"><span class="header-section-number">2.2.2</span> Allele probabilities and frequencies</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Read depth </span></span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a>depth <span class="ot">&lt;-</span> <span class="fu">apply</span>(TAB[,<span class="sc">-</span><span class="fu">c</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">9</span>)], <span class="dv">2</span>, <span class="cf">function</span>(x) <span class="fu">as.integer</span>(<span class="fu">unlist</span>(<span class="fu">lapply</span>(<span class="fu">strsplit</span>(<span class="fu">as.character</span>(x), <span class="at">split =</span> <span class="st">":"</span>), <span class="cf">function</span>(y) y[<span class="dv">2</span>]))))</span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Genotype probabilities, changed by NA for the uncovered sites </span></span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a>gen_prob <span class="ot">&lt;-</span> <span class="fu">apply</span>(TAB[,<span class="sc">-</span><span class="fu">c</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">9</span>)], <span class="dv">2</span>, <span class="cf">function</span>(x) <span class="fu">unlist</span>(<span class="fu">lapply</span>(<span class="fu">strsplit</span>(<span class="fu">as.character</span>(x), <span class="at">split =</span> <span class="st">":"</span>), <span class="cf">function</span>(y) y[<span class="dv">4</span>])))</span>
<span id="cb3-6"><a href="#cb3-6" aria-hidden="true" tabindex="-1"></a>gen_prob[<span class="fu">which</span>(depth<span class="sc">==</span><span class="dv">0</span>)] <span class="ot">&lt;-</span> <span class="cn">NA</span></span>
<span id="cb3-7"><a href="#cb3-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-8"><a href="#cb3-8" aria-hidden="true" tabindex="-1"></a><span class="co"># Proba alternative allele</span></span>
<span id="cb3-9"><a href="#cb3-9" aria-hidden="true" tabindex="-1"></a>altern_proba <span class="ot">&lt;-</span> <span class="fu">apply</span>(gen_prob, <span class="dv">2</span>, <span class="cf">function</span>(x) (<span class="fu">as.numeric</span>(<span class="fu">unlist</span>(<span class="fu">lapply</span>(<span class="fu">strsplit</span>(<span class="fu">as.character</span>(x), <span class="at">split =</span> <span class="st">","</span>), <span class="cf">function</span>(y) y[<span class="dv">2</span>])))<span class="sc">+</span><span class="dv">2</span><span class="sc">*</span><span class="fu">as.numeric</span>(<span class="fu">unlist</span>(<span class="fu">lapply</span>(<span class="fu">strsplit</span>(<span class="fu">as.character</span>(x), <span class="at">split =</span> <span class="st">","</span>), <span class="cf">function</span>(z) z[<span class="dv">3</span>]))))<span class="sc">/</span><span class="dv">2</span>)</span>
<span id="cb3-10"><a href="#cb3-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-11"><a href="#cb3-11" aria-hidden="true" tabindex="-1"></a><span class="co"># Frequency of the alternative allele for each locus and population</span></span>
<span id="cb3-12"><a href="#cb3-12" aria-hidden="true" tabindex="-1"></a>TAB_pop <span class="ot">&lt;-</span> <span class="fu">lapply</span>(<span class="fu">unique</span>(pops),<span class="cf">function</span>(x) altern_proba[,<span class="fu">which</span>(pops<span class="sc">==</span>x)])</span>
<span id="cb3-13"><a href="#cb3-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-14"><a href="#cb3-14" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(TAB_pop) <span class="ot">&lt;-</span> <span class="fu">unique</span>(pops)</span>
<span id="cb3-15"><a href="#cb3-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-16"><a href="#cb3-16" aria-hidden="true" tabindex="-1"></a>freq_pop <span class="ot">&lt;-</span> <span class="fu">lapply</span>(TAB_pop, <span class="cf">function</span>(x) <span class="fu">apply</span>(x, <span class="dv">1</span>, <span class="cf">function</span>(y) <span class="fu">sum</span>(y, <span class="at">na.rm =</span> T)<span class="sc">/</span><span class="fu">sum</span>(<span class="sc">!</span><span class="fu">is.na</span>(y))))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
</section>
<section id="functions-to-optimize-selection" class="level2" data-number="2.3">
<h2 data-number="2.3" class="anchored" data-anchor-id="functions-to-optimize-selection"><span class="header-section-number">2.3</span> Functions to optimize selection</h2>
<section id="select-regions-to-select-sources" class="level3" data-number="2.3.1">
<h3 data-number="2.3.1" class="anchored" data-anchor-id="select-regions-to-select-sources"><span class="header-section-number">2.3.1</span> Select regions to select sources</h3>
<p>Based on the idea of Regional admixture provenancing <em>(citation)</em>, seed sources were selected regionally for each restoration site. Three groups of source populations were subsetted for the three planting sites, removing XVC and HR because of their northern ancestry. More info on the regional ancestry detailed in <em>(citation)</em>.</p>
<div class="tabset-margin-container"></div><div class="panel-tabset">
<ul class="nav nav-tabs" role="tablist"><li class="nav-item" role="presentation"><a class="nav-link active" id="tabset-1-1-tab" data-bs-toggle="tab" data-bs-target="#tabset-1-1" role="tab" aria-controls="tabset-1-1" aria-selected="true">Optimizing genetic diversity</a></li><li class="nav-item" role="presentation"><a class="nav-link" id="tabset-1-2-tab" data-bs-toggle="tab" data-bs-target="#tabset-1-2" role="tab" aria-controls="tabset-1-2" aria-selected="false">Estimate the genetic load</a></li><li class="nav-item" role="presentation"><a class="nav-link" id="tabset-1-3-tab" data-bs-toggle="tab" data-bs-target="#tabset-1-3" role="tab" aria-controls="tabset-1-3" aria-selected="false"><em>N</em> Source selection</a></li></ul>
<div class="tab-content">
<div id="tabset-1-1" class="tab-pane active" role="tabpanel" aria-labelledby="tabset-1-1-tab">
<div class="cell">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="co"># function to estimate allelic richness after rarefaction</span></span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a>rarefy_AR <span class="ot">&lt;-</span> <span class="cf">function</span>(data, g, <span class="at">bootstraping=</span><span class="dv">100</span>){</span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a>  Nijg <span class="ot">&lt;-</span> <span class="fu">list</span>()</span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a>  Njg <span class="ot">&lt;-</span> g<span class="sc">*</span><span class="dv">2</span></span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true" tabindex="-1"></a>  nbind <span class="ot">&lt;-</span> <span class="fu">ncol</span>(data)</span>
<span id="cb4-6"><a href="#cb4-6" aria-hidden="true" tabindex="-1"></a>  Nij <span class="ot">&lt;-</span> <span class="fu">list</span>()</span>
<span id="cb4-7"><a href="#cb4-7" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span>(boot <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span>bootstraping){</span>
<span id="cb4-8"><a href="#cb4-8" aria-hidden="true" tabindex="-1"></a>    inds <span class="ot">&lt;-</span> <span class="fu">sample</span>(<span class="dv">1</span><span class="sc">:</span>nbind, g, <span class="at">replace =</span> <span class="cn">FALSE</span>)</span>
<span id="cb4-9"><a href="#cb4-9" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span>(g<span class="sc">==</span><span class="dv">1</span>){</span>
<span id="cb4-10"><a href="#cb4-10" aria-hidden="true" tabindex="-1"></a>      Nij[[boot]] <span class="ot">&lt;-</span> data[,inds]<span class="sc">*</span><span class="dv">2</span>}</span>
<span id="cb4-11"><a href="#cb4-11" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span>(g<span class="sc">&gt;</span><span class="dv">1</span>){</span>
<span id="cb4-12"><a href="#cb4-12" aria-hidden="true" tabindex="-1"></a>      Nij[[boot]] <span class="ot">&lt;-</span> <span class="fu">apply</span>(data[,inds], <span class="dv">1</span>, <span class="cf">function</span>(x) <span class="fu">sum</span>(x, <span class="at">na.rm =</span> T)<span class="sc">/</span><span class="fu">sum</span>(<span class="sc">!</span><span class="fu">is.na</span>(x)))<span class="sc">*</span>g<span class="sc">*</span><span class="dv">2</span>}</span>
<span id="cb4-13"><a href="#cb4-13" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb4-14"><a href="#cb4-14" aria-hidden="true" tabindex="-1"></a>  Nijg <span class="ot">&lt;-</span> <span class="fu">rowMeans</span>(<span class="fu">do.call</span>(cbind, Nij), <span class="at">na.rm =</span> T)</span>
<span id="cb4-15"><a href="#cb4-15" aria-hidden="true" tabindex="-1"></a>  Qijg <span class="ot">&lt;-</span> (Njg<span class="sc">-</span>Nijg)<span class="sc">/</span>Njg</span>
<span id="cb4-16"><a href="#cb4-16" aria-hidden="true" tabindex="-1"></a>  Pijg <span class="ot">&lt;-</span> <span class="dv">1</span><span class="sc">-</span>Qijg</span>
<span id="cb4-17"><a href="#cb4-17" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(Pijg)</span>
<span id="cb4-18"><a href="#cb4-18" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
<div id="tabset-1-2" class="tab-pane" role="tabpanel" aria-labelledby="tabset-1-2-tab">
<p>The following function is used to estiamte the ratio of nonsynonymous/synonymous mutation based on the annotation from <a href="https://pcingola.github.io/SnpEff/">SnpEff v5.1</a> <em>(Cingolani et al., 2012)</em>, which was used to annotate genetic variants to functional class based on Norway spruce genome annotation. The functional categories viz.&nbsp;missense variant, splice acceptor variant, splice donor variant, splice region variant, start lost, stop gained, stop lost were used to designate as non-synonymous mutation in our calculation for genetic load.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a>genetic_load <span class="ot">&lt;-</span> <span class="cf">function</span>(data, category){</span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a>                nonsyn_sites <span class="ot">&lt;-</span> <span class="fu">which</span>(category<span class="sc">==</span><span class="st">"missense_variant"</span> <span class="sc">|</span> category<span class="sc">==</span><span class="st">"splice_acceptor_variant"</span> <span class="sc">|</span> category<span class="sc">==</span><span class="st">"splice_donor_variant"</span> <span class="sc">|</span> category<span class="sc">==</span><span class="st">"splice_region_variant"</span> <span class="sc">|</span> category<span class="sc">==</span><span class="st">"start_lost"</span> <span class="sc">|</span> category<span class="sc">==</span><span class="st">"stop_gained"</span> <span class="sc">|</span> category<span class="sc">==</span><span class="st">"stop_lost"</span>)</span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a>                freq_nonsyn <span class="ot">&lt;-</span> <span class="fu">mean</span>(data[nonsyn_sites], <span class="at">na.rm =</span> T)</span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a>                freq_syn <span class="ot">&lt;-</span> <span class="fu">mean</span>(data[<span class="sc">-</span>nonsyn_sites], <span class="at">na.rm =</span> T)</span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true" tabindex="-1"></a>                ratio_2 <span class="ot">&lt;-</span> freq_nonsyn<span class="sc">/</span>freq_syn</span>
<span id="cb5-6"><a href="#cb5-6" aria-hidden="true" tabindex="-1"></a>                <span class="fu">return</span>(ratio_2)</span>
<span id="cb5-7"><a href="#cb5-7" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
<div id="tabset-1-3" class="tab-pane" role="tabpanel" aria-labelledby="tabset-1-3-tab">
<p>This function combines the rarefy_AR and genetic_load function to estimate expected heterozygosity (Hexp), allelic richness and genetic load in all combinations of <strong><em>P</em></strong> populations. The <strong><em>P</em></strong> depends on the number of sources one decides to select for their restoration site.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="co"># function to estimate Hexp, Allelic Richness and Genetic Load in all combination of P populations</span></span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a>optimize <span class="ot">&lt;-</span> <span class="cf">function</span>(data, P){</span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true" tabindex="-1"></a>            <span class="co"># Total diversity and load with all the populations</span></span>
<span id="cb6-5"><a href="#cb6-5" aria-hidden="true" tabindex="-1"></a>            TAB_tot <span class="ot">&lt;-</span> <span class="fu">do.call</span>(cbind,data)</span>
<span id="cb6-6"><a href="#cb6-6" aria-hidden="true" tabindex="-1"></a>            freq_tot <span class="ot">&lt;-</span> <span class="fu">apply</span>(TAB_tot, <span class="dv">1</span>, <span class="cf">function</span>(y) <span class="fu">sum</span>(y, <span class="at">na.rm =</span> T)<span class="sc">/</span><span class="fu">sum</span>(<span class="sc">!</span><span class="fu">is.na</span>(y)))</span>
<span id="cb6-7"><a href="#cb6-7" aria-hidden="true" tabindex="-1"></a>            hexp_tot <span class="ot">&lt;-</span> <span class="fu">mean</span>(<span class="dv">2</span><span class="sc">*</span>freq_tot<span class="sc">*</span>(<span class="dv">1</span><span class="sc">-</span>freq_tot), <span class="at">na.rm =</span> T)</span>
<span id="cb6-8"><a href="#cb6-8" aria-hidden="true" tabindex="-1"></a>            <span class="co">#all_rich_tot &lt;- mean(rarefy_AR(TAB_tot, ncol(TAB_tot)), na.rm = T)</span></span>
<span id="cb6-9"><a href="#cb6-9" aria-hidden="true" tabindex="-1"></a>            genetic_load_tot <span class="ot">&lt;-</span> <span class="fu">genetic_load</span>(TAB_tot, category)</span>
<span id="cb6-10"><a href="#cb6-10" aria-hidden="true" tabindex="-1"></a>            </span>
<span id="cb6-11"><a href="#cb6-11" aria-hidden="true" tabindex="-1"></a>            <span class="co"># Genetic diversity and load with only a subset of P populations</span></span>
<span id="cb6-12"><a href="#cb6-12" aria-hidden="true" tabindex="-1"></a>            hexp_sub <span class="ot">&lt;-</span> <span class="fu">list</span>()</span>
<span id="cb6-13"><a href="#cb6-13" aria-hidden="true" tabindex="-1"></a>           </span>
<span id="cb6-14"><a href="#cb6-14" aria-hidden="true" tabindex="-1"></a>            genetic_load_sub <span class="ot">&lt;-</span> <span class="fu">list</span>()</span>
<span id="cb6-15"><a href="#cb6-15" aria-hidden="true" tabindex="-1"></a>            names <span class="ot">&lt;-</span> <span class="fu">list</span>()</span>
<span id="cb6-16"><a href="#cb6-16" aria-hidden="true" tabindex="-1"></a>            comb <span class="ot">&lt;-</span> <span class="fu">combn</span>(<span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(data), P, <span class="at">simplify =</span> F)</span>
<span id="cb6-17"><a href="#cb6-17" aria-hidden="true" tabindex="-1"></a>            <span class="cf">for</span>(i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(comb)){</span>
<span id="cb6-18"><a href="#cb6-18" aria-hidden="true" tabindex="-1"></a>              TAB_sub <span class="ot">&lt;-</span> <span class="fu">do.call</span>(cbind, data[comb[[i]]])</span>
<span id="cb6-19"><a href="#cb6-19" aria-hidden="true" tabindex="-1"></a>              freq_sub <span class="ot">&lt;-</span> <span class="fu">apply</span>(TAB_sub, <span class="dv">1</span>, <span class="cf">function</span>(y) <span class="fu">sum</span>(y, <span class="at">na.rm =</span> T)<span class="sc">/</span><span class="fu">sum</span>(<span class="sc">!</span><span class="fu">is.na</span>(y)))</span>
<span id="cb6-20"><a href="#cb6-20" aria-hidden="true" tabindex="-1"></a>              hexp_sub[i] <span class="ot">&lt;-</span> <span class="fu">mean</span>(<span class="dv">2</span><span class="sc">*</span>freq_sub<span class="sc">*</span>(<span class="dv">1</span><span class="sc">-</span>freq_sub), <span class="at">na.rm =</span> T)</span>
<span id="cb6-21"><a href="#cb6-21" aria-hidden="true" tabindex="-1"></a>              </span>
<span id="cb6-22"><a href="#cb6-22" aria-hidden="true" tabindex="-1"></a>              genetic_load_sub[i] <span class="ot">&lt;-</span> <span class="fu">genetic_load</span>(TAB_sub, category)</span>
<span id="cb6-23"><a href="#cb6-23" aria-hidden="true" tabindex="-1"></a>              names[i] <span class="ot">&lt;-</span> <span class="fu">paste</span>(<span class="fu">names</span>(data[comb[[i]]]), <span class="at">collapse=</span><span class="st">"_"</span>)</span>
<span id="cb6-24"><a href="#cb6-24" aria-hidden="true" tabindex="-1"></a>            }</span>
<span id="cb6-25"><a href="#cb6-25" aria-hidden="true" tabindex="-1"></a>            TAB_sub <span class="ot">&lt;-</span> <span class="fu">do.call</span>(rbind, <span class="fu">lapply</span>(<span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(hexp_sub), <span class="cf">function</span>(x) <span class="fu">c</span>(<span class="at">Hexp =</span> hexp_sub[[x]], <span class="at">GenLoad =</span> genetic_load_sub[[x]]))) <span class="co">#AllRich = all_rich_sub[[x]], </span></span>
<span id="cb6-26"><a href="#cb6-26" aria-hidden="true" tabindex="-1"></a>            TAB <span class="ot">&lt;-</span> <span class="fu">rbind</span>(<span class="fu">c</span>(<span class="at">Hexp =</span> hexp_tot, <span class="at">GenLoad =</span> genetic_load_tot), TAB_sub) <span class="co">#AllRich = all_rich_tot, </span></span>
<span id="cb6-27"><a href="#cb6-27" aria-hidden="true" tabindex="-1"></a>            <span class="fu">rownames</span>(TAB) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"total"</span>, <span class="fu">unlist</span>(names))</span>
<span id="cb6-28"><a href="#cb6-28" aria-hidden="true" tabindex="-1"></a>            </span>
<span id="cb6-29"><a href="#cb6-29" aria-hidden="true" tabindex="-1"></a>            <span class="fu">return</span>(TAB)</span>
<span id="cb6-30"><a href="#cb6-30" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>
</div>
</section>
<section id="apply-the-function-to-get-optimal-source-combinations" class="level3" data-number="2.3.2">
<h3 data-number="2.3.2" class="anchored" data-anchor-id="apply-the-function-to-get-optimal-source-combinations"><span class="header-section-number">2.3.2</span> Apply the function to get optimal source combinations</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Optimization sources for site in Maryland</span></span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a>res_maryland <span class="ot">&lt;-</span> <span class="fu">optimize</span>(TAB_pop_maryland, <span class="dv">3</span>)</span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a><span class="fu">which.max</span>(res_maryland[<span class="sc">-</span><span class="dv">1</span>,<span class="dv">1</span>]<span class="sc">/</span>res_maryland[<span class="sc">-</span><span class="dv">1</span>,<span class="dv">2</span>])</span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true" tabindex="-1"></a>res_maryland[<span class="fu">c</span>(<span class="dv">1</span>,<span class="fu">which.max</span>(res_maryland[<span class="sc">-</span><span class="dv">1</span>,<span class="dv">1</span>]<span class="sc">/</span>res_maryland[<span class="sc">-</span><span class="dv">1</span>,<span class="dv">2</span>])<span class="sc">+</span><span class="dv">1</span>),]</span>
<span id="cb7-5"><a href="#cb7-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-6"><a href="#cb7-6" aria-hidden="true" tabindex="-1"></a><span class="co"># Optimization sources for site in West Virginia</span></span>
<span id="cb7-7"><a href="#cb7-7" aria-hidden="true" tabindex="-1"></a>res_westvirginia_4 <span class="ot">&lt;-</span> <span class="fu">optimize</span>(TAB_pop_westvirginia, <span class="dv">4</span>)</span>
<span id="cb7-8"><a href="#cb7-8" aria-hidden="true" tabindex="-1"></a><span class="fu">which.max</span>(res_westvirginia_4[<span class="sc">-</span><span class="dv">1</span>,<span class="dv">1</span>]<span class="sc">/</span>res_westvirginia_4[<span class="sc">-</span><span class="dv">1</span>,<span class="dv">2</span>])</span>
<span id="cb7-9"><a href="#cb7-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-10"><a href="#cb7-10" aria-hidden="true" tabindex="-1"></a><span class="co"># Optimization sources for site in Virginia</span></span>
<span id="cb7-11"><a href="#cb7-11" aria-hidden="true" tabindex="-1"></a>res_virginia_4 <span class="ot">&lt;-</span> <span class="fu">optimize</span>(TAB_pop_virginia, <span class="dv">4</span>)</span>
<span id="cb7-12"><a href="#cb7-12" aria-hidden="true" tabindex="-1"></a><span class="fu">which.max</span>(res_virginia_4[<span class="sc">-</span><span class="dv">1</span>,<span class="dv">1</span>]<span class="sc">/</span>res_virginia_4[<span class="sc">-</span><span class="dv">1</span>,<span class="dv">2</span>])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
</section>
<section id="factors-affecting-source-selection" class="level2" data-number="2.4">
<h2 data-number="2.4" class="anchored" data-anchor-id="factors-affecting-source-selection"><span class="header-section-number">2.4</span> Factors affecting source selection</h2>
</section>
<section id="saving-data-for-downstream-analysis" class="level2" data-number="2.5">
<h2 data-number="2.5" class="anchored" data-anchor-id="saving-data-for-downstream-analysis"><span class="header-section-number">2.5</span> Saving data for downstream analysis</h2>


</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
<nav class="page-navigation">
  <div class="nav-page nav-page-previous">
      <a href="./intro.html" class="pagination-link">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Introduction</span></span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="./summary.html" class="pagination-link">
        <span class="nav-page-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Summary</span></span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->



</body></html>